Libraries/LibEmail/IMAPClient.cpp

```diff
@@ -0,0 +1,320 @@
+    dbg() << "Step 1: connection: ";
+    if (connected)
+        dbg() << "Connected successfully";
+    else
+        dbg() << "Failed to connect";
+
+    // Setup thread for receiving/queuing server responses
+    m_receive_thread = LibThread::Thread::construct(
+        [this] {
+            while (true) {
+                const StringBuilder response{ m_socket->receive(1000) };
+                if (response.is_empty()) {
+                    // For some reason servers sometimes send empty messages
+                    continue;
+                }
+                m_queue_lock.lock();
+                m_message_queue.append(response.to_string());
+                m_queue_lock.unlock();
+            }
+            return 0;
+        });
+    m_receive_thread->start();
+}
+
+IMAPClient::~IMAPClient()
+{
+    m_receive_thread->quit();
+}
+
+ResponseStatus IMAPClient::login(StringView username, StringView password)
+{
+    StringBuilder command;
+    command.append("login ");
+    command.append(username);
+    command.append(" ");
+    command.append(password);
+    bool status = send_command(command.string_view());
+    if (!status) {
+        dbg() << "Failed to send LOGIN command";
+        return ResponseStatus::FailedToSend;
+    }
+    String response{ receive_response() };
+    dbg() << "Step 2: Login: \n";
+    const auto res_status = get_response_status(response.view());
+    if (res_status == ResponseStatus::Ok) {
+        m_state = ConnectionState::Authenticated;
+        dbg() << "Logged in successfully";
+    } else {
+        dbg() << "Failed to login";
+    }
+    return res_status;
+}
+
+ResponseStatus IMAPClient::select_mailbox(StringView mailbox)
+{
+    if (m_state == ConnectionState::NotAuthenticated) {
+        dbg() << "Cannot select: improper current state";
+        return ResponseStatus::FailedToSend;
+    }
+    StringBuilder command;
+    command.append("select ");
+    command.append(mailbox);
+    bool status = send_command(command.string_view());
+    if (!status) {
+        dbg() << "Failed to send SELECT command";
+        return ResponseStatus::FailedToSend;
+    }
+
+    String response{ receive_response() };
+    dbg() << "Step 3: Selected mailbox: " << mailbox;
+    const auto res_status = get_response_status(response.view());
+    if (res_status == ResponseStatus::Ok) {
+        m_state = ConnectionState::Selected;
+        dbg() << "Selected mailbox successfully";
+    } else {
+        dbg() << "Failed to select mailbox";
+        dbg() << response;
+    }
+    return res_status;
+}
+
+ResponseStatus IMAPClient::create_mailbox(StringView mailbox)
+{
+    if (m_state == ConnectionState::NotAuthenticated) {
+        dbg() << "Cannot create mailbox: improper current state";
+        return ResponseStatus::FailedToSend;
+    }
+
+    StringBuilder command;
+    command.append("create ");
+    command.append(mailbox);
+    bool status = send_command(command.string_view());
```

Comment:
I see this design you make.. and I was wondering if you can improve. 
You have a worker thread, to which you send a message, and then wait until you read a message (`receive_response()`). This will lock up the thread calling.

IMHO, you can design an async API - and to each of these methods, also pass a "callback" - which will be called from the thread you created in the constructor.

Reply:
@elcuco I agree. Unfortunately it will require quite a bit of rearchitecting, so I haven't gotten around to it. The way it is architected now worked decently on macOS natively, but it seems like Serenity supports the asynchronous style a lot more naturally. Thanks for your feedback!